PowerPointLabs/PowerPointLabs/ActionFramework/NarrationsLab/TempStorage.cs

```diff
@@ -0,0 +1,193 @@
+                // due to some error
+                OverwriteFileUsing(presFullName, zipFullPath);
+                ExtractMediaFiles(zipFullPath, tempPath);
+            }
+            catch (Exception e)
+            {
+                ErrorDialogBox.ShowDialog(CommonText.ErrorPrepareMedia, "Files cannot be linked.", e);
+            }
+        }
+
+        private static void OverwriteFileUsing(string presFullName, string zipFullPath)
+        {
+            try
+            {
+                FileDir.DeleteFile(zipFullPath);
+                FileDir.CopyFile(presFullName, zipFullPath);
+            }
+            catch (Exception e)
+            {
+                ErrorDialogBox.ShowDialog(CommonText.ErrorAccessTempFolder, string.Empty, e);
+            }
+        }
+
+        private static void RecreateDirectory(string tempPath)
+        {
+            try
+            {
+                if (Directory.Exists(tempPath))
+                {
+                    Directory.Delete(tempPath, true);
+                }
+            }
+            catch (Exception e)
+            {
+                ErrorDialogBox.ShowDialog(CommonText.ErrorCreateTempFolder, string.Empty, e);
+            }
+            finally
+            {
+                Directory.CreateDirectory(tempPath);
+            }
+        }
+
+        private static void RegulatePresentationName(Presentation pres, string tempPath, ref string presName,
+            ref string presFullName)
+        {
+            // this function is used to handle "embed on other application" issue. In this case,
+            // all of presentation name, path and full name do not match the usual rule: name is
+            // "Untitled", path is empty string and full name is "slide in XX application". We need
+            // to regulate these fields properly.
+
+            presName = presName.AppendIfAbsent(fileSuffix);
+
+            if (tempPath != null)
+            {
+                // every time when recorder pane is open,
+                // save this presentation's copy, which will be used
+                // to load audio files later
+                pres.SaveCopyAs(tempPath + presName);
+                presFullName = tempPath + presName;
+            }
+        }
+
+        private static void ExtractMediaFiles(string zipFullPath, string tempPath)
+        {
+            try
+            {
+                ZipStorer zip = ZipStorer.Open(zipFullPath, FileAccess.Read);
+                List<ZipStorer.ZipFileEntry> dir = zip.ReadCentralDir();
+
+                Regex regex = new Regex(SlideXmlSearchPattern);
+
+                foreach (ZipStorer.ZipFileEntry entry in dir)
+                {
+                    string name = Path.GetFileName(entry.FilenameInZip);
+
+                    if (name == null)
+                    {
+                        continue;
+                    }
+
+                    if (name.Contains(".wav") ||
```

Comment:
Can use optional chaining `name?.Contains`. Can ignore, narrations lab will be discontinued.